doc/developer/arrangements.md

```diff
@@ -60,15 +60,130 @@ Some of these input arrangements could be shared, but most likely the intermedia
+Our `Reduce` operator builds a non-trivial dataflow subgraph.
+
+As of v0.7.1, the dataflow subgraph depends on 1) what types of aggregations are
+involved and 2) the type of input for the reduce.
+
+### Distinct (no aggregations)
+
+Creates two arrangements, each with row count `<number of groups>` and row width
+`<width of key>`.
+
+```
+In -> A -> A -> Out
+```
+
+### Only accumulable aggregations (sum/counts/any/all):
+
+Creates two arrangements, each with row count `<number of groups>` and row width
+`<width of key> + <width of aggregation columns>`.
+
+```
+In -> A -> A -> Out
+```
+
+### Only hierarchical aggregations (min/max) on an append-only input:
+
+Same as the "Only accumulable aggregations" case above.
+
+### Only hierarchical aggregations (min/max) on other inputs:
+
+Hierarchical aggregations on inputs that are not append-only have
+`ceil(log_16(expected_group_size))` stages, which are intended to filter the
+number of candidate rows by a factor of 16 per stage. If an
+`expected_group_size` has not been supplied via query hint, the default value
+is 4 billion.
+
+Each stage involves two arrangements.
+The first arrangement of the final stage is supposed to contain the top 16 rows
+for each group.
+The first arrangement of the second to last stage is supposed to contain the top
+16^2 = 256 rows for each group.
+The first arrangement of the third to last stage is supposed to contain the top
+16^3 = 4096 rows for each group.
+... and so on. Note that the first arrangement of the first stage will always
+contain all rows from the input.
+
+The second arrangement of the final stage is supposed to contain the top row
+for each group.
+The second arrangement of the second to last stage is supposed to contain the
+rows contained in the first arrangement of the second to last stage but not
+contained in the first arrangement of the final stage.
+The second arrangement of the third to last stage is supposed to contain the
+rows contained in the first arrangement of the third to last stage but not
+contained in the first arrangement of the second to last stage.
+... and so on.
+
+A two-stage dataflow would look like this:
+```
+In -> PrevStgArr1 -> PrevStgArr2 -> Subtract -> FinStgArr1 -> FinStgArr2 -> Out
+           \                           /
+            -------------------------->
+```
+
+The row width for all arrangements in this case are
+`<width of key> + <width of aggregation columns>`.
+
+A general heuristic for the total row count of all the arrangements, assuming a
+well-chosen `expected_group_size`, is
+`(2 + epsilon) * <number of distinct rows>`.
+The first arrangement of the first stage will always contain all rows from the
+input. `<row count of the first arrangement of the second stage> + `
+`<row count of the second arrangement of the first stage>` is always equal to
+`<row count of the first arrangement of the first stage>`.
+
+Specifying an `expected_group_size` that is magnitudes larger than the actual
```

Comment:
Probably also worth mentioning that if you undershoot `expected_group_size` by a lot then we would expect the update latency to go up for some input streams